### DISCUSSION TITLE
How to use actor-query-process-annotate-source-binding

### DISCUSSION THREAD

**Discussion by balhoff** [comunica/comunica#1396] [Q&A] [answered]
I came across [this actor](https://github.com/comunica/comunica/blob/master/packages/actor-query-process-annotate-source-binding/README.md), and it would be really useful to me to report the data sources for query solutions. I tried adding it to https://github.com/comunica/comunica/blob/master/engines/config-query-sparql/config/query-process/actors.json and found that it did start outputting `_source` bindings, but the value is always `[]`, e.g.:

```json
{"cell":"http://purl.obolibrary.org/obo/CL_0000182","cell_label":"\"hepatocyte\"","_source":"\"[]\""}
```

Is there further configuration needed to populate the `_source` binding?

**Comment by rubensworks** [comment]
@RubenEschauzier Could you answer this one?

**Comment by RubenEschauzier** [comment, marked as answer]
The `actor-query-process-annotate-source-binding` actor is only the last step to get the provenance of solutions from Comunica. You need to add a few more actors to compute the provenance information.

This [PR](https://github.com/comunica/comunica-feature-link-traversal/pull/138/files#diff-40b42573405aea18fe5dca764e3163acec46b99a947c9beef2add7e4e25b964f) aims to add these configs to Comunica link-traversal.

In summary, you need to add the following `query-source-identify-hypermedia` actor to your config:
```
    {
      "@id": "urn:comunica:default:query-source-identify-hypermedia/actors#annotate-source",
      "@type": "ActorQuerySourceIdentifyHypermediaAnnotateSource",
      "mediatorMergeBindingsContext": { "@id": "urn:comunica:default:merge-bindings-context/mediators#main" },
      "mediatorQuerySourceIdentifyHypermedia": { "@id": "urn:comunica:default:query-source-identify-hypermedia/mediators#main" },
      "beforeActors": [
        { "@id": "urn:comunica:default:query-source-identify-hypermedia/actors#none" },
        { "@id": "urn:comunica:default:rdf-resolve-quad-pattern/actors#sparql" },
        { "@id": "urn:comunica:default:query-source-identify-hypermedia/actors#qpf" }
      ]
    }
```
This will add the source of the bindings to their context.

Then to ensure that during the join of bindings, the sources required for the result are the union of the two input binding sources, you need to add the following `merge-bindings-context` actor:

```
{
      "@id": "urn:comunica:default:merge-bindings-context/actors#source-binding-union",
      "@type": "ActorMergeBindingsContextUnion",
      "contextKey": "@comunica/bus-merge-bindings-context:sourceBinding"
}
```
Then the `actor-query-process-annotate-source-binding` actor will take the source information from the binding context and put it as an actual binding in the query result.

**Reply by balhoff** [reply]
Thank you for the info, @RubenEschauzier! I haven't quite gotten it to work, but my repo is kind of a mess and I'll start from a clean slate and try again. Do you expect this to work for a federation of TPF and SPARQL endpoints?

**Reply by RubenEschauzier** [reply]
No problem, let me know if any issues show up in the clean slate.
I would expect it to work on both TPF and SPARQL endpoints, but I haven't tested it specifically on that combination, so do let me know if any errors or unwanted behaviors appear.

**Reply by balhoff** [reply]
@RubenEschauzier here are the changes I made to try this out: https://github.com/comunica/comunica/compare/master...balhoff:comunica:test-annotate-source

I added some logging to see if the new actors were running. When I run a query against two QPF endpoints, I get these lines printed (you can see where I put `console.log` in the diff):

```
ActorQueryProcessAnnotateSourceBinding
ActorMergeBindingsContextUnion
ActorMergeBindingsContextUnion
ActorQuerySourceIdentifyHypermediaAnnotateSource - TEST
ActorMergeBindingsContextUnion
ActorMergeBindingsContextUnion
ActorQuerySourceIdentifyHypermediaAnnotateSource - TEST
ActorMergeBindingsContextUnion
ActorMergeBindingsContextUnion
ActorMergeBindingsContextUnion
```

The `run` method of `ActorQuerySourceIdentifyHypermediaAnnotateSource` never seems to be called, and the `_source` binding is not populated. Am I missing something? Thank you!

**Reply by RubenEschauzier** [reply]
This might be a bug that occurs when annotating QPF endpoints. Are the QPF endpoints available publicly so I can reproduce the problem locally?
(Note I'll be on leave again from 13-25 August, so responses will be somewhat delayed in that period)

**Reply by balhoff** [reply]
@RubenEschauzier yes they are public, here is an example query:

```
node bin/query.js 'https://frink.apps.renci.org/ldf/wikidata' 'https://frink.apps.renci.org/ldf/ubergraph' 'PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX wdtn: <http://www.wikidata.org/prop/direct-normalized/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX cell: <http://purl.obolibrary.org/obo/CL_0000000>
PREFIX abdomen: <http://purl.obolibrary.org/obo/UBERON_0000916>
PREFIX part_of: <http://purl.obolibrary.org/obo/BFO_0000050>
SELECT DISTINCT ?cell ?cell_label ?article ?title
WHERE {
  ?cell rdfs:subClassOf cell: .
  ?cell part_of: abdomen: .
  ?wikicell wdtn:P7963 ?cell .
  ?article wdt:P921 ?wikicell .
  ?article rdfs:label ?title .
  ?cell rdfs:label ?cell_label .
}
LIMIT 20'
```

No hurry! And thank you for all your help.

**Reply by RubenEschauzier** [reply]
Update: The problem is a bug with `ActorQuerySourceIdentifyHypermediaAnnotateSource`. As usual, fixing this revealed another bug. Both should be fixed in #1420, which (hopefully) only requires slight polishing before being merge-ready.
